internal/helpers/msgraph/application.go

```diff
@@ -0,0 +1,386 @@
+		}
+	}
+
+	// then set the new roles
+	properties := msgraph.Application{
+		ID:       application.ID,
+		AppRoles: newRoles,
+	}
+
+	if _, err := client.Update(ctx, properties); err != nil {
+		return fmt.Errorf("setting App Roles for Application with object ID %q: %+v", *application.ID, err)
+	}
+
+	return nil
+}
+
+func ApplicationSetOAuth2PermissionScopes(ctx context.Context, client *msgraph.ApplicationsClient, application *msgraph.Application, newScopes *[]msgraph.PermissionScope) error {
+	if application.ID == nil {
+		return fmt.Errorf("Cannot use Application model with nil ID")
+	}
+
+	if newScopes == nil {
+		newScopes = &[]msgraph.PermissionScope{}
+	}
+
+	// OAuth2 Permission Scopes must be disabled before they can be edited or removed.
+	// Since we cannot match them by ID, we have to disable all the scopes, and replace them in one pass.
+	app, status, err := client.Get(ctx, *application.ID)
+	if err != nil {
+		if status == http.StatusNotFound {
+			return fmt.Errorf("application with ID %q was not found", *application.ID)
+		}
+
+		return fmt.Errorf("retrieving Application with object ID %q: %+v", *application.ID, err)
+	}
+
+	// don't update if no changes to be made
+	if app.Api != nil && app.Api.OAuth2PermissionScopes != nil && reflect.DeepEqual(*app.Api.OAuth2PermissionScopes, *newScopes) {
+		return nil
+	}
+
+	// first disable any existing scopes
+	if app.Api != nil && app.Api.OAuth2PermissionScopes != nil && len(*app.Api.OAuth2PermissionScopes) > 0 {
+		properties := msgraph.Application{
+			ID: application.ID,
+			Api: &msgraph.ApplicationApi{
+				OAuth2PermissionScopes: app.Api.OAuth2PermissionScopes,
+			},
+		}
+
+		for _, scope := range *properties.Api.OAuth2PermissionScopes {
+			*scope.IsEnabled = false
+		}
+
+		if _, err := client.Update(ctx, properties); err != nil {
+			return fmt.Errorf("disabling OAuth2 Permission Scopes for Application with object ID %q: %+v", *application.ID, err)
+		}
+	}
+
+	// then set the new scopes
+	properties := msgraph.Application{
+		ID: application.ID,
+		Api: &msgraph.ApplicationApi{
+			OAuth2PermissionScopes: newScopes,
+		},
+	}
+
+	if _, err := client.Update(ctx, properties); err != nil {
+		return fmt.Errorf("setting OAuth2 Permission Scopes for Application with object ID %q: %+v", *application.ID, err)
+	}
+
+	return nil
+}
+
+func ApplicationSetOwners(ctx context.Context, client *msgraph.ApplicationsClient, application *msgraph.Application, desiredOwners []string) error {
+	if application.ID == nil {
+		return fmt.Errorf("Cannot use Application model with nil ID")
+	}
+
+	owners, _, err := client.ListOwners(ctx, *application.ID)
+	if err != nil {
+		return fmt.Errorf("retrieving owners for Application with object ID %q: %+v", *application.ID, err)
+	}
+
+	existingOwners := *owners
+	ownersForRemoval := utils.Difference(existingOwners, desiredOwners)
+	ownersToAdd := utils.Difference(desiredOwners, existingOwners)
+
+	if ownersForRemoval != nil {
```

Comment:
Since you can't delete the last owner, this will fail if you are changing a single owner. Recommend adding owners first and moving this block underneath the `ownersToAdd` block so there is always at least one owner.

Reply:
@Yanson Great catch, thanks. I've updated so that owners for groups and applications are both added before old ones are removed.

Reply:
Thanks for updating. Is there any way to add a test to prevent this regressing?

> ... and have no additional permissions

Unfortunately in our organisation we are not permitted such elevated permission and may only mange the groups we created (and own).

Also, I don't think it's related to this PR, but the msgraph API - I was unable to remove the second-to-last owner if the last owner is a Service Principal. If you know anything about that, it would be great to address.

Reply:
> Thanks for updating. Is there any way to add a test to prevent this regressing?

Not easily since we typically use Global Admin to separate concerns between permissions of our test account and the accounts it manages during testing, but I'll look into it.

> Also, I don't think it's related to this PR, but the msgraph API - I was unable to remove the second-to-last owner if the last owner is a Service Principal. If you know anything about that, it would be great to address.

Interesting, I haven't been bitten by this but will do some testing 👍